pandas/core/algorithms.py

```diff
@@ -785,16 +785,67 @@ def duplicated(values, keep='first'):
           occurrence.
         - ``last`` : Mark duplicates as ``True`` except for the last
           occurrence.
-        - False : Mark all duplicates as ``True``.
+        - False : Mark all duplicates as ``True``. This option is not
+          compatible with ``return_inverse``.
+    return_inverse : boolean, default False
+        If True, also return the selection of (integer) indices from the array
+        of unique values (created e.g. by selecting the boolean complement of
+        the first output, or by using `.drop_duplicates` with the same
+        `keep`-parameter) that can be used to reconstruct "values".
+
+        .. versionadded:: 0.24.0
 
     Returns
     -------
-    duplicated : ndarray
+    duplicated : ndarray or tuple of ndarray if ``return_inverse`` is True
     """
 
+    if return_inverse and keep is False:
+        raise ValueError("The parameters return_inverse=True and "
+                         "keep=False cannot be used together (impossible "
+                         "to calculate an inverse when discarding all "
+                         "instances of a duplicate).")
+
     values, dtype, ndtype = _ensure_data(values)
     f = getattr(htable, "duplicated_{dtype}".format(dtype=ndtype))
-    return f(values, keep=keep)
+    isdup = f(values, keep=keep)
+    if not return_inverse:
+        return isdup
+    elif not isdup.any():
+        # no need to calculate inverse if no duplicates
+        inv = np.arange(len(values))
+        return isdup, inv
+
+    if keep == 'first':
+        # o2u: original indices to indices of ARRAY of unique values
+        # u2o: reduplication from array of unique values to original array
```

Comment:
I'd rather you not use ``np.unique`` at all, its not as performant as ``pd.unique``, doesn't handle all dtypes, and sorts. 

Further pls, pls use actual names here, and really avoid using abbreviations in any library code.

Reply:
@jreback 
> I'd rather you not use `np.unique` at all, its not as performant as `pd.unique`, doesn't handle all dtypes, and sorts.

Yes, it would be nicer to have this implemented in the cython hashtable functions, but that performance improvement is for a follow-up. `np.unique` is an easy solution (and we're only calling it on a series of ints, not objects, because the factorization for that hasn't changed!).

> Further pls, pls use actual names here, and really avoid using abbreviations in any library code.

There's a fully commented, thoroughly explained and very localized part where these appear. Not sure how this is unclear, but will adapt...

Reply:
Finally, the core changes are minimal (but I understand that it looks like a lot). TLDR: Implementation moves to `core.algorithms`, everything else wraps around that (+ doc improvements). Tests are expanded, and ASVs added.